go/storage/mkvs/db/badger/badger.go

```diff
@@ -672,6 +686,56 @@ func (d *badgerNodeDB) Prune(ctx context.Context, version uint64) error {
 	return nil
 }
 
+func (d *badgerNodeDB) StartMultipartInsert() error {
+	if d.multipart {
+		return api.ErrMultipartInProgress
+	}
+
+	d.metaUpdateLock.Lock()
+	defer d.metaUpdateLock.Unlock()
+
+	txn := d.db.NewTransactionAt(tsMetadata, false)
+	defer txn.Discard()
+
+	opt := badger.DefaultIteratorOptions
+	opt.Prefix = multipartRestoreNodeLogKeyFmt.Encode()
+	it := txn.NewIterator(opt)
+	defer it.Close()
+	for it.Rewind(); it.Valid(); it.Next() {
+		return api.ErrMultipartInProgress
+	}
+	return nil
+}
+
+func (d *badgerNodeDB) FinishMultipartInsert(removeNodes bool) error {
+	d.multipart = false
+	d.metaUpdateLock.Lock()
+	defer d.metaUpdateLock.Unlock()
+	txn := d.db.NewTransactionAt(tsMetadata, true)
+	defer txn.Discard()
+
+	opts := badger.DefaultIteratorOptions
+	opts.Prefix = multipartRestoreNodeLogKeyFmt.Encode()
+	it := txn.NewIterator(opts)
+	defer it.Close()
+	for it.Rewind(); it.Valid(); it.Next() {
+		key := it.Item().Key()
+		if removeNodes {
+			var hash hash.Hash
+			if !multipartRestoreNodeLogKeyFmt.Decode(key, &hash) {
+				panic("mkvs/badger: bad iterator")
+			}
+			if err := txn.Delete(nodeKeyFmt.Encode(&hash)); err != nil {
```

Comment:
Consider using a write batch (see `WriteBatch`) for deletion otherwise this needs to handle going over the transaction size limit (e.g., `ErrTxnTooBig`).